### DISCUSSION TITLE
Full examples of modifying part of a file with chezmoi

### DISCUSSION THREAD

**Discussion by jmlucjav** [twpayne/chezmoi#3996] [Q&A] [answered]
Hi,

I am trying to use the 'modify_' approach here https://www.chezmoi.io/user-guide/manage-different-types-of-file/ to handle just part of a file. But I am not fully understanding what files I need to to do 'chezmoi add' for, the original file (let's call it x.rc) or the modify_x.rc, or both?

There is even an issue https://github.com/twpayne/chezmoi/discussions/1746 asking about an example, but it's not clear what file needs to be added either.

much appreciated!

**Comment by twpayne** [comment]
You need to create the `modify_` script manually in your source directory. You can't use `chezmoi add` for this.

**Comment by jmlucjav** [comment]
thanks Tom, couple follow up questions...

- x.rc is 'chezmoi add'-ed as usual?
- so if I use chezmoi in 3 computers, I need to take care to keep the modify_ script in sync in the source dir manually, is that it?

**Reply by bradenhilton** [reply]
> x.rc is 'chezmoi add'-ed as usual?

`x.rc` is the target file for both `x.rc` and `modify_x.rc`. You can't have multiple files in the source state referring to the same target.

`x.rc`'s contents will be set to the output of the `modify_` script, even if the output is empty. If you don't want `x.rc` to ever be empty, you need to ensure that the script outputs a sufficient minimal configuration.

> so if I use chezmoi in 3 computers, I need to take care to keep the modify_ script in sync in the source dir manually, is that it?

`modify_x.rc` is the only file that is needed. If you need different contents on these machines then you need to use templating.

Here is a contrived example:

`~/data.json`:

```json
{
  "color": "red"
}
```

`<chezmoi source>/modify_data.json`:

```text
// chezmoi:modify-template

{{- $currentConfig := .chezmoi.stdin | fromJson -}}
{{- $modifiedConfig := .chezmoi.stdin | fromJson -}}

{{- $_ := set $modifiedConfig "color" "blue" -}}

{{- if deepEqual $currentConfig $modifiedConfig -}}
{{-   .chezmoi.stdin }}
{{- else -}}
{{-   $modifiedConfig | toPrettyJson }}
{{- end -}}
```

```diff
❯ chezmoi diff
diff --git a/data.json b/data.json
index 2f67018551056b7c2e17a2c6da5b7ecbe7a0a15c..c7f2e0cfc95110175c59e9579dae0fc1e3e92c0c 100666
--- a/data.json
+++ b/data.json
@@ -1,3 +1,3 @@
 {
-  "color": "red"
+  "color": "blue"
 }
```

**Reply by jmlucjav** [reply]
thanks for the explanation...bear with me, I must be missing something obvious cause still not working for me...

> > x.rc is 'chezmoi add'-ed as usual?
> 
> `x.rc` is the target file for both `x.rc` and `modify_x.rc`. You can't have multiple files in the source state referring to the same target.
> 

ok got it, I created ~/.local/share/chezmoi/dot_config/modify_kdiff3rc:

```
#!/bin/bash

# copy the data to modify to a temporary file
tmpfile=$(mktemp)
trap "rm -f ${tmpfile}" EXIT
cat > "${tmpfile}"

#remove unwanted lines
sed -i '/shell-geometry/d' ${tmpfile}

# write the data to stdout
cat "${tmpfile}"
```

tested it invoking it and it does what I want (discard changes in line containing 'shell-geometry'). I push it without using 'chezmoi add' 

`chezmoi git add dot_config/modify_kdiff3rc && chezmoi git -- commit -m 'modify_kdiff3rc' && chezmoi git push`

Made sure  ~/.local/share/chezmoi/dot_config/kdiff3rc does not exist anymore.

But now, if I change a buch of lines in ~/.config/kdiff3rc (so I want those changes propagated to other host when I do chezmoi update), nothing happens, my changes stay just in this host. 

Sorry about this, I'm quite obtuse when it comes to git and chezmoi feels like black magic sometimes

**Reply by bradenhilton** [reply, marked as answer]
When using a `modify_` script, the contents of the file are no longer managed in the "traditional" way by chezmoi. Put simply, chezmoi just feeds the current contents of the file to the script via `stdin`, and uses its `stdout` to set the new contents of the file.

If you have changes you want to propagate, this would need to be done with the script itself, not the file.

You *might* want to explore using template functions such as `fromIni`, `toIni`, `setValueAtPath`, `deleteValueAtPath` etc. if you require something more robust than simple pattern replacements.

**Reply by jmlucjav** [reply]
> When using a `modify_` script, the contents of the file are no longer managed in the "traditional" way by chezmoi. Put simply, chezmoi just feeds the current contents of the file to the script via `stdin`, and uses its `stdout` to set the new contents of the file.
> 
> If you have changes you want to propagate, this would need to be done with the script itself, not the file.

oh ok...that is a big caveat that defeats the purpose of using modify_....

> 
> You _might_ want to explore using template functions such as `fromIni`, `toIni`, `setValueAtPath`, `deleteValueAtPath` etc. if you require something more robust than simple pattern replacements.

I will look into those and see if I can get it working without modify_ thanks!

**Reply by bradenhilton** [reply]
> that is a big caveat that defeats the purpose of using modify_....

I'm not sure I understand what you mean by this. Can you give a concrete example of what you want to achieve, and why you think a `modify_` script isn't suitable?

**Reply by jmlucjav** [reply]
Take kdiff3rc for example, this is an .ini type file, and has maybe 100 entries. I am interested in synching this file over all my hosts, but there is one entry, 'shell-geometry' that is about the window size etc. Depending on my screen in each host this value changes every time, and I don't want this change to 'pollute' my git repo, I am not interested in propagating this value to other hosts, but I want all other changes. 

I understood (maybe wrong) that I cannot achieve that with modify_ (getting changes from one host to the others)

**Reply by twpayne** [reply]
If you're modifying complex `.ini` files, check out [`chezmoi_modify_manager`](https://vorpalblade.github.io/chezmoi_modify_manager/).

**Reply by jmlucjav** [reply]
I did try it, but it was more complex to understand than chezmoi itself, so I dropped it

**Comment by jmlucjav** [comment]
Update:

In case this is useful for someone. I think I found I solution that works well for me. I wanted:

- the target file to be kept in chezmoi, propagating to all hosts (this was fundamental)
- ignore some lines from that file (like typical timestamp values etc).

I added a hook pre re-add:

```
[hooks.re-add.pre]
    command = ".../.chezmoiscripts/pre_add_hook_discard_lines.sh"
```

And in the sh I just use sed to remove the lines I want to discard (of course other modifications would be possible).

Maybe would be worth mentioning this option here https://www.chezmoi.io/user-guide/manage-different-types-of-file/#manage-part-but-not-all-of-a-file ?
